crates/circuit/src/circuit_instruction.rs

```diff
@@ -163,7 +163,7 @@ impl CircuitInstruction {
         qubits: Option<&Bound<PyAny>>,
         clbits: Option<&Bound<PyAny>>,
         params: SmallVec<[Param; 3]>,
-        label: Option<String>,
+        label: Option<&str>,
```

Comment:
I think this particular one is mistaken; it's correct that the `label` is owned in this constructor.

Reply:
My intention was to move the creation of the `String` into the constructor rather than the caller. Is the problem that this method `py_new` is called from Python and has to build a `String` rather than `&str`?  Then a ref is passed and another `String` is created with `label.map(|x| x.to_string())` which is of course wasteful ?

Reply:
What I'm thinking of is that if the `label` is `Some`, then we're always going to need to get an owned object.  There's no situation in which we pass `Some` to the `label` and we _won't_ need to clone it to get an owned copy, so there's nothing gained by allowing it to accept a ref, and there's something potentially lost if the caller had to create the label specifically (say from a `format!`) but can't give over the ownership.

I don't know 100% what PyO3 does when it's got an incoming `Bound<'py, PyString>` (i.e. if it makes a `String` or directly a `&str` - I suspect the latter), but accepting an owned `String` allows all possible callers to use the most efficient form.